File: Baekjoon/Problem_1000~1999/Problem_1021/Problem_1021.py
import sys
from collections import deque

input = lambda: sys.stdin.readline().rstrip()
get_first = lambda queue: queue.popleft()
move_left = lambda queue: queue.append(queue.popleft())
move_right = lambda queue: queue.appendleft(queue.pop())

def find(lst, target):
    l, r = 0, 0
    length = len(lst)
    
    # left
    cur = 0
    while lst[cur] != target:
        l += 1
        cur = (cur - 1) % length
        
    # right
    cur = 0
    while lst[cur] != target:
        r += 1
        cur = (cur + 1) % length
    
    return l, r
    
if __name__ == "__main__":
    N, M = map(int, input().split()) # 큐의 크기 N, 뽑아내려는 수의 개수 M
    
    queue = deque(range(1, N+1))
    
    answer = 0
    for e in list(map(int, input().split())):
        
        l, r = find(queue, e)
        
        if l <= r:
            answer += l
            for i in range(l): move_right(queue)
            
        else:
            answer += r
            for i in range(r):move_left(queue)
            
        get_first(queue)
    
    print(answer)

File: Baekjoon/Problem_1000~1999/Problem_1021/Problem_1021.cpp
#include <cstdio>
#include <deque>
#include <utility>

using namespace std;

pair<int, int> find(const deque<int> &deq, const int target) {
    int l = 0, r = 0;
    int length = deq.size();

    // left
    int cur = 0;
    while (deq[cur] != target) {
        l += 1;
        cur = cur == 0 ? length - 1 : (cur - 1) % length;
    }

    // right
    cur = 0;
    while (deq[cur] != target) {
        r += 1;
        cur = (cur + 1) % length;
    }

    return {l, r};
}

int main() {
    int N, M; scanf("%d %d", &N, &M);

    deque<int> deq;
    for (int i = 1; i <= N; i++) deq.push_back(i);

    int answer = 0;
    for (int i = 0; i < M; i++) {
        int target; scanf("%d", &target);

        pair<int, int> tmp = find(deq, target);
        int l = tmp.first, r = tmp.second;

        if (l <= r) {
            answer += l;

            for (int j = 0; j < l; j++) {
                deq.push_front(deq.back());
                deq.pop_back();
            }
        } else {
            answer += r;
            
            for (int j = 0; j < r; j++) {
                deq.push_back(deq.front());
                deq.pop_front();
            }
        }

        deq.pop_front();
    }

    printf("%d\n", answer);

    return 0;
}